FEATURE: ``TYPO3.TypoScript:Debug`` TypoScript-Object

mficzel opened the pull request:
A TypoScript Object "Debug" is added as an equivalent to the ``f:debug`` view helper.
The Object will display the value of all child-keys in a debug window.

```
tsPath = TYPO3.TypoScript:Debug {
    # optional: set title for the debug output
    # title = 'Debug'
  
    # optional: show result as plaintext
    # plaintext = TRUE

    # If only the value-key is given it is debugged directly,
    # otherwise all keys except title an plaintext are debugged.
    value = "hello neos world"

    # Additional values for debugging
    documentTitle = ${q(documentNode).property('title')}
    documentPath = ${documentNode.path}
}
```

aertmann commented:
hey nice one, been dreaming of this for a long time.. although I thought about an Eel helper for it.. maybe both would make sense?

it needs to be documented though

aertmann commented:
also it could be nice if it extended Array, so you could do something like if `value` is empty/undefined

    Debug {
        test = 1
        nodes = ${q(node).children().get()}
    }

mficzel commented:
Would'nt the documentation be updated automatically. As far as i know the part inside References is autocreated.

mficzel commented:
I do'nt get the point in extending Array. You already can do this:
`
foo = Debug {
  value = RawArray {
    test  = 1
    node = ${q(node) ..} 
  }
}
`
What would be the exact benefit in extending Array?

aertmann commented:
ah good point, it should.. but would still be good to mention in the documentation and not just the reference

aertmann commented:
lazyness, I find writing value = RawArray redundant/unnecessary

mficzel commented:
Laziness is a good point :-). I personally dislike that in that case all properties except 'title' and 'plaintext' would be debugged. So an array with some extra-special keys.

Another downside would be that you always debug a dictionary with a key 'value' and not a simple value.

mficzel commented:
I added a section for Inside TypoScript

aertmann commented:
`title` would still be outputted so easy to spot and `plaintext` is pretty unusual so should be okay

aertmann commented:
regarding "always debug a dictionary with a key 'value' and not a simple value", that not what I suggested.. if you provide value then that is debugged and not the array

mficzel commented:
Hmmm ... that could be done but i still do'nt like that the behavior changes completely by adding the magic key "value". The other TS-Objects do'nt behave like this and i would favor consistency over convenience. Imagine the TYPO3.TypoScriopt:Value Object would behave like this.

mficzel commented:
I rebased the code on the current master and implemented the array behavior @aertmann suggested. The current implementation will only display value if that is the only key. If other keys except title and plaintext are defined an array of all those keys is debugged.

dimaip commented on TYPO3.Neos/Documentation/CreatingASite/TypoScript/InsideTypoScript.rst:
```diff
@@ -501,6 +501,28 @@ a property using the ``@if`` meta-property::
+		value = "hello neos world"
+
+		# Additional values for debugging
+		document_title = ${q(documentNode).property('title')}
```
I know it's just an example, but best to change it to upperCamelCase, i.e. `documentTitle`, because that's against our style.

dimaip commented:
I tried it out and it works as expected, so :+1: from me.

mficzel commented:
Adjusted the variable Names in the pr and the documentation to camelCase like @dimaip suggested.

aertmann commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/TypoScriptObjects/DebugImplementation.php:
```diff
@@ -0,0 +1,78 @@
+            $debugData[$key] = $this->tsValue($key);
+        }
+
+        if (count($debugData) == 0) {
```
should be `===`

dimaip replied:
I wonder why StyleCI didn't pick it up? I stopped looking at such things after we set up this linter.

aertmann commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/TypoScriptObjects/DebugImplementation.php:
```diff
@@ -0,0 +1,78 @@
+
+        if (count($debugData) == 0) {
+            $debugData = null;
+        } elseif (array_key_exists('value', $debugData) && count($debugData) == 1) {
```
same here

mficzel commented:
Replaced == in comparisons with === like @aertmann suggested.

mficzel commented:
I have no idea where the test failures complaining that the "registerMockObject" method is not defined are coming from but i doubt that my code has anything to do with that.

mficzel commented:
After the recent fixes in the ci the tests are green again \o/
